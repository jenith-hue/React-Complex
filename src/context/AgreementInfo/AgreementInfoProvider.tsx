import { CancelTokenSource } from 'axios';
import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useRef,
} from 'react';
import { useLocation } from 'react-router-dom';
import { getCancelTokenSource } from '../../api/client';
import { getAgreementInfo } from '../../api/Customer';
import {
  AgreementInfoResponse,
  AgreementInfoStore,
  CustomerLocationState,
} from '../../types/types';

export interface AgreementInfoState {
  stores: Partial<AgreementInfoStore>[];
  total: number;
  numberOfLoadedAgreements: number;
  hasApiError: boolean;
  loading: boolean;
}

export interface AgreementInfoDispatchState {
  fetchAgreementInfo: () => Promise<void | AgreementInfoResponse>;
}

export const AgreementInfoStateContext = createContext<AgreementInfoState>(
  {} as AgreementInfoState
);

export const AgreementInfoDispatchContext =
  createContext<AgreementInfoDispatchState>({} as AgreementInfoDispatchState);

export const AgreementInfoProvider = (props: { children: ReactNode }) => {
  const [stores, setStores] = React.useState<AgreementInfoStore[]>([]);
  const [total, setTotal] = React.useState<number>(0);
  const [numberOfLoadedAgreements, setNumberOfLoadedAgreements] =
    React.useState<number>(0);
  const [hasApiError, setHasApiError] = React.useState(false);
  const [loading, setLoading] = React.useState(false);
  const location = useLocation<CustomerLocationState>();
  const initialCallToken = useRef<CancelTokenSource>();

  useEffect(() => {
    return () => {
      if (initialCallToken?.current) {
        initialCallToken.current?.cancel();
      }
    };
  }, []);

  const fetchAgreementInfo = async () => {
    if (initialCallToken?.current) {
      initialCallToken.current?.cancel();
    }
    initialCallToken.current = getCancelTokenSource();

    const customerId =
      location?.state?.customer?.customerId ||
      location?.pathname?.split('/')[3];

    if (!customerId) return;
    setLoading(true);
    setHasApiError(false);
    getAgreementInfo(customerId, initialCallToken.current?.token)
      .then((response: any) => {
        const { total, stores: storesFromApi } = response;
        const numberOfLoadedAgreementsUntilNow = storesFromApi.reduce(
          (partialSum: any, storeObj: any) => {
            return (storeObj?.agreements?.length || 0) + partialSum;
          },
          numberOfLoadedAgreements
        );
        setTotal(total);
        setNumberOfLoadedAgreements(numberOfLoadedAgreementsUntilNow);
        setStores([...stores, ...storesFromApi]);
      })
      .catch(() => {
        setHasApiError(true);
      })
      .finally(() => setLoading(false));
  };

  return (
    <AgreementInfoStateContext.Provider
      value={{
        stores,
        total,
        numberOfLoadedAgreements,
        hasApiError,
        loading,
      }}
    >
      <AgreementInfoDispatchContext.Provider
        value={{
          fetchAgreementInfo,
        }}
      >
        {props.children}
      </AgreementInfoDispatchContext.Provider>
    </AgreementInfoStateContext.Provider>
  );
};

export const useAgreementInfo = () => useContext(AgreementInfoStateContext);

export const useAgreementInfoActions = () =>
  useContext(AgreementInfoDispatchContext);
