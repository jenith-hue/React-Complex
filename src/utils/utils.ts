import { AxiosResponse } from 'axios';
import { add, sub, format } from 'date-fns';

import { isEmpty } from 'lodash';
import {
  ALL_OPTION,
  CONTENT_TYPE,
  DATE_FORMAT,
  Language,
  STORE_NUMBER_KEY,
} from '../constants/constants';
import { ReferenceItem, RouteResponse } from '../types/types';
import * as crypto from 'crypto';

export const sanitizeURL = (url: string): string => {
  if (!url) {
    return '';
  }

  const badURLRegex = RegExp('^((https)|(http)):/{3,}');
  const isBadURL = badURLRegex.test(url);

  if (isBadURL) {
    return 'https://' + url.replace(badURLRegex, '');
  }

  return url;
};

export const abbrev = (strToCutOff: string): string => {
  if (strToCutOff && strToCutOff.length > 13) {
    return strToCutOff.substr(0, 13) + '...';
  }

  return strToCutOff;
};

export const getErrorMessage = (response?: AxiosResponse<any>) => {
  if (response?.status === 400 && response?.data.errors[0].code) {
    return response?.data.errors[0].message;
  }

  return null;
};

export const orderByField = (fieldName: string, arrayToOrder: any[]) => {
  if (!arrayToOrder || !arrayToOrder.length || !fieldName) return arrayToOrder;
  return arrayToOrder.sort((a, b) => (a[fieldName] > b[fieldName] ? 1 : -1));
};

export const orderByFieldAndValue = (
  value: string,
  fieldName: string,
  arrayToOrder: any[]
) => {
  if (!arrayToOrder || !arrayToOrder.length || !fieldName) return arrayToOrder;
  return arrayToOrder.sort((a) => (a[fieldName] === value ? -1 : 1));
};

export const mapReferenceResponse = (references: ReferenceItem[]) => {
  if (!references || !references.length) return references;
  return references.map((reference) => ({
    label: reference.description,
    value: reference.referenceCode,
    ...reference,
  }));
};

// Not deleting method, because might be useful for Route selection in customer info header.
export const mapRoutesResponse = (routes: RouteResponse[]) => {
  if (!routes || !routes.length) return routes;
  return routes.map((routeObj) => ({
    label: routeObj.routeDescription,
    value: routeObj.routeCode,
    ...routeObj,
  }));
};

export const addAllOption = (arrayWithOptions: any[]) => {
  if (!arrayWithOptions) {
    return [
      {
        label: ALL_OPTION,
        value: ALL_OPTION,
      },
    ];
  }

  return [
    {
      label: ALL_OPTION,
      value: ALL_OPTION,
    },
    ...arrayWithOptions,
  ];
};

export const pipe =
  (...fns: any) =>
  (x: any) =>
    fns.reduce((v: any, f: (value: any) => void) => f(v), x);

export const formatDate = (date: Date, dateFormat = DATE_FORMAT): string => {
  if (isValidDate(date)) return format(date, dateFormat);
  return '';
};

// Dates are stored with local timezone of the respective store.
// We want to show the same data, by ignoring the user's local date settings
// e.g. 2021-07-19T16:07:42 or 2021-07-20
// yyyy-mm-dd
export const formatDateString = (
  dateAsString: string | undefined,
  returnedFormat?: string
): string => {
  if (!dateAsString || typeof dateAsString !== 'string') {
    return '';
  }
  let dateWithoutHours;
  if (dateAsString.includes('T')) {
    dateWithoutHours = dateAsString.split('T')[0];
  } else {
    dateWithoutHours = dateAsString;
  }

  // [0] = 'yyyy'; [1] = 'mm'; [2] = 'dd'
  let dateAsArray = [] as string[];
  if (dateWithoutHours.includes('-')) {
    dateAsArray = dateWithoutHours.split('-');
  } else if (dateWithoutHours.includes('/')) {
    dateAsArray = dateWithoutHours.split('/');
  }

  if (dateAsArray?.length !== 3) {
    return '';
  }

  if (returnedFormat === Language.Spanish) {
    return `${dateAsArray[2]} de ${spanishMonths[dateAsArray[1]]} de ${
      dateAsArray[0]
    }`;
  }

  if (returnedFormat === 'MM-dd-yyyy') {
    return `${dateAsArray[1]}-${dateAsArray[2]}-${dateAsArray[0]}`;
  }
  // mm/dd/yyyy
  return `${dateAsArray[1]}/${dateAsArray[2]}/${dateAsArray[0]}`;
};

export const toLocaleStringWithMinutes = (date: Date): string => {
  if (isValidDate(date))
    return date.toLocaleString('en-US', {
      hour: 'numeric',
      minute: 'numeric',
      hour12: true,
    });

  return '';
};

// e.g. "2022-03-31T02:56:44.738-05:00"
export const formatStringDateHoursAndMinutes = (
  dateAsString: string | undefined
): string => {
  try {
    if (dateAsString && isValidDate(new Date(dateAsString))) {
      const split1 = dateAsString.split('T');
      if (!split1[1]) {
        return '';
      }
      const split2 = split1[1].split(':');
      if (!split2[1]) {
        return '';
      }

      const hours = Number(split2[0]);
      const minutes = split2[1];

      let hoursToDisplay = hours % 12;
      let meridian = 'AM';

      if (hours >= 12) {
        meridian = 'PM';
      }
      if (hoursToDisplay === 0) {
        hoursToDisplay = 12;
      }

      return `${hoursToDisplay}:${minutes} ${meridian}`;
    }

    return '';
  } catch {
    return '';
  }
};

export type ListStatus =
  | 'initial'
  | 'loading'
  | 'apiError'
  | 'empty'
  | 'success';
export interface ComponentStateProps {
  loading?: boolean;
  hasApiError?: boolean;
  response?: any;
}

export const getComponentState = ({
  loading,
  hasApiError,
  response,
}: ComponentStateProps): ListStatus => {
  if (loading) return 'loading';
  if (hasApiError) return 'apiError';

  if (response) {
    if (Array.isArray(response)) {
      return response.length > 0 ? 'success' : 'empty';
    } else if (isEmpty(response)) {
      return 'empty';
    }
    return 'success';
  } else {
    return 'initial';
  }
};

export const DAYS_OF_WEEK = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];

export function isValidDate(d: any) {
  return d instanceof Date && !isNaN(d.getTime());
}

export const moveItemToTheEnd = (
  items: any,
  fieldName: string,
  value: string | number
): any => {
  if (!fieldName) return items;
  if (!value) return items;
  if (!Array.isArray(items)) return items;
  if (items.length < 2) return items;
  const itemsCopy = [...items];

  const index = itemsCopy.findIndex((item) => item[fieldName] === value);
  if (index < 0) return items;

  itemsCopy.splice(index, 1);
  return [...itemsCopy, items[index]];
};

export const formatPhoneNumberForPrintLetter = (phone: string) => {
  const cleaned = ('' + phone).replace(/\D/g, '');
  const match = cleaned.match(/^(\d{3})(\d{3})(\d{4})$/);

  if (match) {
    return '(' + match[1] + ') ' + match[2] + '-' + match[3];
  }

  return null;
};

export const getLatestItem = (items?: any): any => {
  if (!Array.isArray(items)) return undefined;
  if (items.length > 0) {
    return items[items.length - 1];
  }

  return undefined;
};

export const formatPhoneNumber = (phone: string, isMasked?: boolean) => {
  if (phone && isMasked) return `(***) ***-****`;
  if (!phone || phone.includes('-')) return phone;

  return `(${phone.slice(0, 3)}) ${phone.slice(3, 6)}-${phone.slice(6)}`;
};

export const getSelectedStore = () =>
  sessionStorage.getItem(STORE_NUMBER_KEY) || '';

export const randomNumericStringGenerator = (numberOfDigits = 8) => {
  if (typeof numberOfDigits !== 'number') {
    throw new TypeError(
      `Expected number instead got: ${typeof numberOfDigits}`
    );
  }

  return crypto
    .randomBytes(4)
    .readUInt32BE(0)
    .toString()
    .substr(0, numberOfDigits);
};

export const b64toBlob = (
  b64Data: any,
  contentType = 'application/pdf',
  sliceSize = 512
) => {
  const byteCharacters = atob(b64Data);
  const byteArrays = [];

  for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
    const slice = byteCharacters.slice(offset, offset + sliceSize);

    const byteNumbers = new Array(slice.length);
    for (let i = 0; i < slice.length; i++) {
      byteNumbers[i] = slice.charCodeAt(i);
    }

    const byteArray = new Uint8Array(byteNumbers);
    byteArrays.push(byteArray);
  }

  return new Blob(byteArrays, { type: contentType });
};

export const formatMoney = (amount: string | number) => {
  if (amount === null || amount === undefined || isNaN(amount as number)) {
    return '$0.00';
  }

  return `$${Number(amount).toFixed(2)}`;
};

export const getContentType = (fileName: string, fileContentType?: string) => {
  if (fileContentType) return fileContentType;
  let result;

  if (fileName && typeof fileName === 'string') {
    const splits = fileName.split('.');
    const extension = splits[splits.length - 1] as string;
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    result = CONTENT_TYPE[extension];
  }

  return result || CONTENT_TYPE.pdf;
};

export const SIX_MONTHS_AGO = sub(new Date(), { months: 6 });
export const SIX_MONTHS_AFTER = add(new Date(), { months: 6 });

export const thirtyDaysBeforeToDate = (toDate: Date) =>
  sub(toDate, { days: 31 });
export const thirtyDaysAfterFromDate = (fromDate: Date) =>
  add(fromDate, { days: 31 });

export const spanishMonths: any = {
  '01': 'enero',
  '02': 'febrero',
  '03': 'marzo',
  '04': 'abril',
  '05': 'mayo',
  '06': 'junio',
  '07': 'julio',
  '08': 'agosto',
  '09': 'septiembre',
  '10': 'octubre',
  '11': 'noviembre',
  '12': 'diciembre',
};
