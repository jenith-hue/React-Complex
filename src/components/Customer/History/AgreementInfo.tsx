import {
  makeStyles,
  RACCOLOR,
  RACTable,
  RACTableCell,
  RACTableRow,
  Typography,
} from '@rentacenter/racstrap';
import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import InfiniteScroll from 'react-infinite-scroll-component';
import {
  ApiStateWrapper,
  Loader,
} from '../../../common/ApiStateWrapper/ApiStateWrapper';
import {
  useAgreementInfo,
  useAgreementInfoActions,
} from '../../../context/AgreementInfo/AgreementInfoProvider';
import { formatDateString } from '../../../utils/utils';
import { AgreementInfoDaysLate } from '../../../types/types';
import { AppRoute } from '../../../config/route-config';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';

const useClasses = makeStyles((theme: any) => ({
  tableContentColor: {
    color: RACCOLOR.INDEPENDENCE,
  },
  tableRoot: {
    marginTop: theme.typography.pxToRem(16),
  },
  // aligns the content in the cell
  storeRowContentCell: {
    display: 'flex',
  },
  storeRowContent: {
    position: 'absolute',
  },
  notVisible: {
    visibility: 'hidden',
  },
  tableBody: {
    position: 'relative',
  },
  tableHead: {
    position: 'sticky',
    top: 0,
    zIndex: 2,
    background: 'white',
  },
  contentHeight: {
    height: theme.typography.pxToRem(400),
  },
  greenColumn: {
    backgroundColor: '#81c783',
  },
  number: {
    color: RACCOLOR.BLUE_CRAYOLA,
    cursor: 'pointer',
  },
}));

export const EMPTY_DAYS_LATE_HISTORY = [
  { dayslate: 'On_Time', count: 0 },
  { dayslate: '1-6_Days_Late', count: 0 },
  { dayslate: '7-14_Days_Late', count: 0 },
  { dayslate: '15+_Days_Late', count: 0 },
];

export const populateDaysLateHistory = (
  daysLateHistory: AgreementInfoDaysLate[] | null | undefined
): AgreementInfoDaysLate[] => {
  return EMPTY_DAYS_LATE_HISTORY.map((emptyDaysLateObj) => {
    const daysLateObjFromAPI = daysLateHistory?.find(
      (obj) => obj.dayslate === emptyDaysLateObj.dayslate
    );

    return daysLateObjFromAPI || emptyDaysLateObj;
  });
};

export const AgreementInfo = () => {
  const { stores, total, numberOfLoadedAgreements, hasApiError, loading } =
    useAgreementInfo();
  const { customerDetails } = useCustomerDetails();
  const { customerId } = customerDetails || {};

  const renderTableHead = () => (
    <>
      <RACTableCell>
        <Typography variant="h5">Agreement</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Due Date</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Type</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">Current</Typography>
      </RACTableCell>
      <RACTableCell color="green">
        <Typography className={classes.greenColumn} variant="h5">
          On Time
        </Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">1-6 Days Late</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">7-14 Days Late</Typography>
      </RACTableCell>
      <RACTableCell>
        <Typography variant="h5">15+ Days Late</Typography>
      </RACTableCell>
    </>
  );

  const daysLateHistoryCellGenerator = (
    daysLateHistory: AgreementInfoDaysLate[]
  ) => {
    const populatedDaysLateHistory = populateDaysLateHistory(daysLateHistory);

    return populatedDaysLateHistory.map(
      ({ count, dayslate }: AgreementInfoDaysLate) => (
        <RACTableCell key={dayslate}>
          <Typography variant="body2" className={classes.tableContentColor}>
            {count}
          </Typography>
        </RACTableCell>
      )
    );
  };
  const renderTableContent = () => (
    <>
      {stores?.map((storeObj) => {
        if (!storeObj?.agreements?.length) return null;

        return storeObj?.agreements?.map(
          (
            {
              agreementNumber,
              agreementId,
              nextDueDate,
              currentDaysLate,
              daysLateHistory,
              contractType,
            },
            index
          ) => {
            const dueDateString = `${formatDateString(nextDueDate)}`;
            const showAgreementNumLink = contractType === 'CLUB' ? false : true;
            return (
              <>
                {index === 0 && (
                  <RACTableRow
                    key={`${index}-${storeObj.storeNumber}`}
                    backgroundColor="white"
                  >
                    <RACTableCell className={classes.storeRowContentCell}>
                      <div className={classes.storeRowContent}>
                        <Typography variant="h4">{`${storeObj.storeNumber} - ${storeObj.city}, ${storeObj.state}`}</Typography>
                      </div>
                    </RACTableCell>
                    <RACTableCell>
                      <Typography
                        noWrap
                        display="inline"
                        variant="body2"
                        className={classes.notVisible}
                      >
                        {/* Using typography to ensure that all rows have the same
                        height(this is not visible in the table)*/}
                        0
                      </Typography>
                    </RACTableCell>
                  </RACTableRow>
                )}
                <RACTableRow key={index} backgroundColor="white">
                  <RACTableCell classes={{ root: classes.number }}>
                    {showAgreementNumLink ? (
                      <Link
                        style={{ color: 'inherit', textDecoration: 'unset' }}
                        to={{
                          pathname: `${AppRoute.CustomerAggrement}/${customerId}/${agreementId}`,
                          search: `?origin=am-customer`,
                        }}
                        className={'disi'}
                      >
                        {agreementNumber}
                      </Link>
                    ) : (
                      <Typography
                        display="inline"
                        variant="body2"
                        className={classes.tableContentColor}
                      >
                        {agreementNumber}
                      </Typography>
                    )}
                  </RACTableCell>
                  <RACTableCell>
                    <Typography
                      display="inline"
                      variant="body2"
                      className={classes.tableContentColor}
                    >
                      {dueDateString}
                    </Typography>
                  </RACTableCell>
                  <RACTableCell>
                    <Typography
                      display="inline"
                      variant="body2"
                      className={classes.tableContentColor}
                    >
                      {contractType}
                    </Typography>
                  </RACTableCell>
                  <RACTableCell>
                    <Typography
                      display="inline"
                      variant="body2"
                      className={classes.tableContentColor}
                    >
                      {currentDaysLate}
                    </Typography>
                  </RACTableCell>
                  {daysLateHistoryCellGenerator(daysLateHistory)}
                </RACTableRow>
              </>
            );
          }
        );
      })}
    </>
  );
  const classes = useClasses();
  const { fetchAgreementInfo } = useAgreementInfoActions();

  const initialLoad = () => {
    fetchAgreementInfo();
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(initialLoad, []);
  return (
    <InfiniteScroll
      next={fetchAgreementInfo}
      dataLength={numberOfLoadedAgreements}
      hasMore={numberOfLoadedAgreements < total}
      height={330}
      loader={<Loader />}
    >
      <ApiStateWrapper
        loading={loading}
        hasApiError={hasApiError}
        response={stores}
        successContent={
          <RACTable
            bodyClasses={{ root: classes.tableBody }}
            headClasses={{ root: classes.tableHead }}
            classes={{ root: classes.tableRoot }}
            renderTableHead={renderTableHead}
            renderTableContent={renderTableContent}
          />
        }
        classes={{
          loader: classes.contentHeight,
          apiError: classes.contentHeight,
          noItems: classes.contentHeight,
        }}
      />
    </InfiniteScroll>
  );
};
