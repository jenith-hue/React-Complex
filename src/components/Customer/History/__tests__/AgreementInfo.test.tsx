/* eslint-disable sonarjs/no-duplicate-string */
/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { AgreementInfo, populateDaysLateHistory } from '../AgreementInfo';
import * as Utils from '../../../../utils/utils';

interface PropWithChildren {
  children: JSX.Element;
}

interface MockRACTableProps {
  renderTableHead: () => JSX.Element;
  renderTableContent: () => JSX.Element;
}
// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '',
  }),
}));

jest.mock(
  'react-infinite-scroll-component',
  () =>
    ({ children, ...rest }: PropWithChildren) =>
      (
        <div data-testid="InfiniteScroll" {...rest}>
          {children}
        </div>
      )
);

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACTableCell: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableCell" {...rest}>
      {children}
    </div>
  ),
  RACTableRow: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableRow" {...rest}>
      {children}
    </div>
  ),
  RACCOLOR: {
    INDEPENDENCE: 'INDEPENDENCE',
  },
  RACTable: ({ renderTableHead, renderTableContent }: MockRACTableProps) => (
    <div data-testid="RACTable">
      <div id="table-head">{renderTableHead()}</div>
      <div id="table-content">{renderTableContent()}</div>
    </div>
  ),
  makeStyles: () => () => ({
    tableContentColor: 'tableContentColor',
    tableRoot: 'tableRoot',
    storeRowContentCell: 'storeRowContentCell',
    storeRowContent: 'storeRowContent',
    notVisible: 'notVisible',
  }),
}));

jest.mock('../../../../context/AgreementInfo/AgreementInfoProvider', () => ({
  useAgreementInfoActions: () => ({
    fetchAgreementInfo: jest.fn(),
  }),
  useAgreementInfo: () => ({
    stores: [
      {
        storeNumber: jest.fn(),
        city: 'Flagstaff',
        state: 'AZ',
        agreements: [
          {
            agreementNumber: '08767673445',
            nextDueDate: '2021-04-26',
            currentDaysLate: '28',
            daysLateHistory: [
              {
                dayslate: '1',
                count: '0',
              },
              {
                dayslate: '2',
                count: '1',
              },
              {
                dayslate: '3',
                count: '0',
              },
              {
                dayslate: '4',
                count: '0',
              },
              {
                dayslate: '5',
                count: '2',
              },
              {
                dayslate: '6',
                count: '0',
              },
              {
                dayslate: '7',
                count: '0',
              },
              {
                dayslate: '8',
                count: '0',
              },
              {
                dayslate: '9',
                count: '0',
              },
              {
                dayslate: '10',
                count: '0',
              },
            ],
          },
        ],
      },
    ],
  }),
}));

describe('AgreementInfo', () => {
  it('renders correctly', () => {
    jest
      .spyOn(Utils, 'formatDateString')
      .mockImplementation(() => '12/12/2021');

    const tree = renderer.create(
      <BrowserRouter>
        <AgreementInfo />
      </BrowserRouter>
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
  describe('populateDaysLateHistory', () => {
    it('populates correctly when all data present', () => {
      const daysLateHistory = [
        { dayslate: 'On_Time', count: 1 },
        { dayslate: '1-6_Days_Late', count: 2 },
        { dayslate: '7-14_Days_Late', count: 3 },
        { dayslate: '15+_Days_Late', count: 4 },
      ];

      const result = populateDaysLateHistory(daysLateHistory);

      expect(result).toStrictEqual([
        { dayslate: 'On_Time', count: 1 },
        { dayslate: '1-6_Days_Late', count: 2 },
        { dayslate: '7-14_Days_Late', count: 3 },
        { dayslate: '15+_Days_Late', count: 4 },
      ]);
    });
    it('populates correctly when all data present (order different)', () => {
      const daysLateHistory = [
        { dayslate: '15+_Days_Late', count: 4 },
        { dayslate: '7-14_Days_Late', count: 3 },
        { dayslate: '1-6_Days_Late', count: 2 },
        { dayslate: 'On_Time', count: 1 },
      ];

      const result = populateDaysLateHistory(daysLateHistory);

      expect(result).toStrictEqual([
        { dayslate: 'On_Time', count: 1 },
        { dayslate: '1-6_Days_Late', count: 2 },
        { dayslate: '7-14_Days_Late', count: 3 },
        { dayslate: '15+_Days_Late', count: 4 },
      ]);
    });
    it('populates correctly when all data missing', () => {
      const daysLateHistory = null;

      const result = populateDaysLateHistory(daysLateHistory);

      expect(result).toStrictEqual([
        { dayslate: 'On_Time', count: 0 },
        { dayslate: '1-6_Days_Late', count: 0 },
        { dayslate: '7-14_Days_Late', count: 0 },
        { dayslate: '15+_Days_Late', count: 0 },
      ]);
    });
    it('populates correctly when data is empty arr', () => {
      const daysLateHistory = [] as any;

      const result = populateDaysLateHistory(daysLateHistory);

      expect(result).toStrictEqual([
        { dayslate: 'On_Time', count: 0 },
        { dayslate: '1-6_Days_Late', count: 0 },
        { dayslate: '7-14_Days_Late', count: 0 },
        { dayslate: '15+_Days_Late', count: 0 },
      ]);
    });
    it('populates correctly when some data missing', () => {
      const daysLateHistory = [
        { dayslate: 'On_Time', count: 1 },
        { dayslate: '7-14_Days_Late', count: 3 },
      ];

      const result = populateDaysLateHistory(daysLateHistory);

      expect(result).toStrictEqual([
        { dayslate: 'On_Time', count: 1 },
        { dayslate: '1-6_Days_Late', count: 0 },
        { dayslate: '7-14_Days_Late', count: 3 },
        { dayslate: '15+_Days_Late', count: 0 },
      ]);
    });
  });
});
