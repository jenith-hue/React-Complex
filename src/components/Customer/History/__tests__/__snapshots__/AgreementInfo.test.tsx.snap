// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgreementInfo renders correctly 1`] = `
<div
  data-testid="InfiniteScroll"
  hasMore={false}
  height={330}
  loader={<Loader />}
  next={[MockFunction]}
>
  <div
    data-testid="RACTable"
  >
    <div
      id="table-head"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          Agreement
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          Due Date
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          Type
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          Current
        </div>
      </div>
      <div
        color="green"
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          On Time
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          1-6 Days Late
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          7-14 Days Late
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="h5"
        >
          15+ Days Late
        </div>
      </div>
    </div>
    <div
      id="table-content"
    >
      <div
        backgroundColor="white"
        data-testid="RACTableRow"
      >
        <div
          className="storeRowContentCell"
          data-testid="RACTableCell"
        >
          <div
            className="storeRowContent"
          >
            <div
              data-testid="Typography"
              variant="h4"
            >
              function () {
        return fn.apply(this, arguments);
      } - Flagstaff, AZ
            </div>
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="notVisible"
            data-testid="Typography"
            display="inline"
            noWrap={true}
            variant="body2"
          >
            0
          </div>
        </div>
      </div>
      <div
        backgroundColor="white"
        data-testid="RACTableRow"
      >
        <div
          classes={
            Object {
              "root": undefined,
            }
          }
          data-testid="RACTableCell"
        >
          <a
            className="disi"
            href="/agreement/info/details/undefined/undefined?origin=am-customer"
            onClick={[Function]}
            style={
              Object {
                "color": "inherit",
                "textDecoration": "unset",
              }
            }
          >
            08767673445
          </a>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            display="inline"
            variant="body2"
          >
            12/12/2021
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            display="inline"
            variant="body2"
          />
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            display="inline"
            variant="body2"
          >
            28
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            variant="body2"
          >
            0
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            variant="body2"
          >
            0
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            variant="body2"
          >
            0
          </div>
        </div>
        <div
          data-testid="RACTableCell"
        >
          <div
            className="tableContentColor"
            data-testid="Typography"
            variant="body2"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
